core/authn.go

```diff
@@ -34,26 +36,54 @@ type tokenResult struct {
 
 func (a *apiAuthn) handler(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-		err := a.auth(req)
-		if err != nil {
-			errorFormatter.Write(req.Context(), rw, err)
+		token, err0 := a.tokenAuth(req)
+		ctx := req.Context()
+		if err0 == nil && token != "" {
+			// if this request was successfully authenticated, pass the token along
+			ctx = context.WithValue(ctx, "token", token)
```

Comment:
Check out the sample code in https://godoc.org/context#Context (on the Value method) and https://godoc.org/context#WithValue. I think we should try to follow that pattern: making an unexported type just for the context key guarantees it won't conflict with any other package.

So here it could be something like

```go
type authnKey int

var (
	tokenKey = iota
	localhostKey
)
```

and this line would be

```go
ctx = context.WithValue(ctx, tokenKey, token)
```